### PR TITLE
Add LLaVA audio (sound) model support

### LINKED ISSUE #4588: Nemotron 3 Nano Omni Training support
Tracking issue for training support of [Nemotron-3-Nano-Omni-30B-A3B](https://huggingface.co/nvidia/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-BF16), a 30B-A3B MoE multimodal model that jointly processes image, video, audio, and text.

## Status

- **Day 0 support already available in Megatron-Bridge.** Checkpoint conversion, SFT, and LoRA recipes are on the [`nemotron_3_omni`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/tree/nemotron_3_omni) branch — see the [examples README](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/nemotron_3_omni/examples/models/vlm/nemotron_3_omni/README.md) for the full walkthrough.
- **Upstream Megatron-LM support is in progress** via #4402 (LLaVA audio/sound modality with Parakeet/FastConformer encoder, dynamic-resolution RADIO, and `nemotron6-moe` plumbing).
- The next regular Megatron-Core release (**0.18**) will include this support out of the box. (0.17 has already been cut, so the changes from #4402 land in 0.18.)

## Links

- PR: https://github.com/NVIDIA/Megatron-LM/pull/4402
- Megatron-Bridge Day-0 branch: https://github.com/NVIDIA-NeMo/Megatron-Bridge/tree/nemotron_3_omni
- Megatron-Bridge example: https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/nemotron_3_omni/examples/models/vlm/nemotron_3_omni/README.md
- Megatron-Bridge PR: TBD

### PR DESCRIPTION
**Pull request by cuichenx**
Adds end-to-end support for an audio/sound modality in `LLaVAModel`, plus the helpers needed to use it at scale.

## Summary of changes

- `megatron/core/models/huggingface/fastconformer_model.py` (new): `ParakeetHuggingFaceModel` wrapper that loads either a NeMo Parakeet checkpoint (`nemo://...`) or an upstream Hugging Face FastConformer model (`hf://...`) via `transformers.AutoModel` / `AutoFeatureExtractor`, with optional gradient checkpointing. Returns `(hidden_states, lengths)` to match `LLaVAModel`'s call site.
- `megatron/core/models/huggingface/module.py`: route `nemo://...parakeet...` and `hf://...parakeet...` paths to `ParakeetHuggingFaceModel` from `get_hf_model_type` / `build_hf_model`. Non-parakeet `nemo://` schemes raise a clear `NotImplementedError`.
- `megatron/core/models/multimodal/llava_model.py`:
  - New `sound_model` / `sound_projection` / `sound_token_index` plumbing, `DEFAULT_SOUND_TOKEN_INDEX`, `SOUND_TOKEN`, `freeze_sound_model` / `freeze_sound_projection`, and corresponding wiring through `forward` / token combination in `_preprocess_data`.
  - `dynamic_resolution` plus RADIO knobs (`force_eval_mode`, `force_cpe_eval_mode`, `interpolate_only_cpe`, `cpe_aspect_ratio_select`, `disable_cpe`, `temporal_patch_dim`, `separate_video_embedder`, `temporal_ckpt_compat`).
  - Extend the SP/CP gating and language-model dispatch to also recognize `nemotron6-moe`; keep using the upstream `HybridModel` / `hybrid_stack_spec`.
  - Extend `pixel_shuffle` with optional `h, w` kwargs for non-square dynamic-resolution patch grids.
- `megatron/core/models/multimodal/context_parallel.py`: add CP utilities for dynamic-resolution vision inputs (split / gather across CP ranks, tubelet-aware split points, padding for unequal media counts, FP8 recipe-aware tail padding).
- `megatron/core/models/vision/radio.py`: dynamic-resolution support; temporal patch dim with optional separate video embedder; CPE controls (force-eval, interpolate-only, aspect-ratio select, disable); parameterized `interpolate_align_corners` / `grid_sample_align_corners`; state-dict pre-hooks to upgrade 2D embedder weights to 3D for temporal compression / from a non-temporal checkpoint.

## Tests

Four new unit-test files (~1.7 k LOC), validated 84 / 84 passing on a 2-rank cw-dfw run:

- `tests/unit_tests/models/test_fastconformer_model.py`: HF backend dtype propagation, sampling-rate plumbing, `gradient_checkpointing`, error paths; NeMo singleton cache keyed by model id; `parakeet` dispatch in `module.py` (positive + negative cases).
- `tests/unit_tests/models/test_llava_sound.py`: `SOUND_TOKEN` / `DEFAULT_SOUND_TOKEN_INDEX` constants; `freeze` honouring the new flags; `has_sounds` sentinel decision logic; sound replacement in `_preprocess_data` for both `sound_pad_to_clip_duration` branches.
- `tests/unit_tests/models/test_multimodal_context_parallel.py`: `_compute_tubelet_aware_split_points` invariants; `_split_num_frames` clipping; `split_to_context_parallel_ranks_dynamic_res` (non-temporal, tubelet-aware, padded, hidden-dim assertion); `gather_from_context_parallel_ranks` with `global_pad`; `GatherFromContextParallelRanks` autograd backward (CP gradient correctness regression guard).
- `tests/unit_tests/models/test_radio_model.py`: constructor / forward / state-dict round-trip; `_state_dict_pre_hook_init_embedder` + `_state_dict_pre_hook_init_video_embedder`; `train()` override with `force_eval_mode`; `pixel_shuffle` square + non-square; constructor coverage for the new RADIO knobs; `_apply_temporal_grouping` direct unit tests.

# What does this PR do ?

End-to-end audio modality for `LLaVAModel` (Parakeet/FastConformer encoder + projection + sound-token replacement) plus dynamic-resolution / temporal-patch support for RADIO and the CP utilities needed for both.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/models/vision/radio.py [hunk 1]
```diff
@@ -359,18 +698,13 @@ def window_select(pos_embed):
...
     if not "vision_model.class_token" in state_dict:
         return
 
-    pad = 32 if module.config.fp8_recipe == "mxfp8" else 16
-
     class_token = state_dict["vision_model.class_token"]
-    if class_token.shape[0] % pad != 0:
-        pad_len = pad - (class_token.shape[0] % pad)
+    if class_token.shape[0] % 16 != 0:
+        pad_len = 16 - (class_token.shape[0] % 16)
```
Comment (cuichenx) [inline megatron/core/models/vision/radio.py]: pad should not be hardcoded to 16, what should be the condition?

File: megatron/core/models/vision/radio.py [hunk 2]
```diff
@@ -124,23 +151,62 @@ def __init__(
...
+                    self._state_dict_pre_hook_init_video_embedder
+                )
+            else:
+                self.register_load_state_dict_pre_hook(self._state_dict_pre_hook_init_embedder)
 
         self.model_type = ModelType.encoder_or_decoder
 
         self.ln_pre = None
         self.ln_post = None
-        self.pg_collection = pg_collection
```
Comment (trintamaki) [inline megatron/core/models/vision/radio.py:142]: should we keep the pg_collection and vp_stage?

File: megatron/core/models/vision/radio.py [hunk 3]
```diff
@@ -161,86 +227,320 @@ def __init__(
...
+                x,
+                'b c (py yy) (px xx) -> b (py px) (c yy xx)',
+                py=py,
+                yy=self.patch_dim,
+                px=px,
+                xx=self.patch_dim,
+            )
 
-        if not HAVE_EINOPS:
-            raise ImportError(
```
Comment (trintamaki) [inline megatron/core/models/vision/radio.py:191]: do we still want to keep checking for this import?

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4402/commits/b373b26c2ac89179c80663f23e0f6be8192b29be

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4402/commits/b6b1d039b870c34523ec39e696dddd694a5ddfb6

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4402/commits/6091e22de20b4b3edae5661a3cf3ee9545497353

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4402/commits/03d8a571cf7119628158f04eddb95f3580ca4b73

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4402/commits/f75d38c91dc6b98be9adb75cab8d0e6d85a4960d

**Review by trintamaki — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
